core/src/main/java/ai/timefold/solver/core/impl/bavet/uni/ForEachExcludingUnassignedUniNode.java

```diff
@@ -26,22 +26,23 @@ public void insert(A a) {
 
     @Override
     public void update(A a) {
-        UniTuple<A> tuple = tupleMap.get(a);
+        var tuple = tupleMap.get(a);
         if (tuple == null) { // The tuple was never inserted because it did not pass the filter.
             insert(a);
         } else if (filter.test(a)) {
-            innerUpdate(a, tuple);
-        } else {
-            super.retract(a); // Call super.retract() to avoid testing the filter again.
+            updateExisting(a, tuple);
+        } else { // Tuple no longer passes the filter.
+            retract(a);
         }
     }
 
     @Override
     public void retract(A a) {
-        if (!filter.test(a)) { // The tuple was never inserted because it did not pass the filter.
+        var tuple = tupleMap.remove(a);
```

Comment:
Why we don't need to filter before removing?

Reply:
Performance optimization.
Removing from a map will return null if the thing was never inserted; therefore I can avoid calling the filter, and still get the same result.